core/commands/pin.go

```diff
@@ -90,32 +90,31 @@ var addPinCmd = &cmds.Command{
 		v := new(dag.ProgressTracker)
 		ctx := v.DeriveContext(req.Context())
 
-		ch := make(chan []*cid.Cid)
+		type pinResult struct {
+			pins []*cid.Cid
+			err  error
+		}
```

Comment:
Huh, I did not know you could define types in functions. Neat!

Reply:
You want to have some *real* fun?

```go
ch := make(chan struct{pins []*cid.Cid, err error}, 1)
//...
ch <- struct{pins []*cid.Cid, err error}{err: errors.New("😎")
```

Reply:
I have some other trick for you:

```go
type System struct {
    cntrSafe struct {
        sync.Mutex
        n int
    }
    mapSafe struct {
        sync.Mutex
        m map[int]int
    }
}
````